clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/names/EcsServerGroupNameResolver.java

```diff
@@ -0,0 +1,142 @@
+
+    // 2. Get the details of the services in the same server group
+    List<List<String>> serviceBatches = Lists.partition(allServices, 10);
+    for (List<String> serviceBatch : serviceBatches) {
+      DescribeServicesRequest request =
+          new DescribeServicesRequest()
+              .withCluster(ecsClusterName)
+              .withServices(serviceBatch)
+              .withInclude("TAGS");
+      DescribeServicesResult result = ecs.describeServices(request);
+      for (Service service : result.getServices()) {
+        Moniker moniker = naming.deriveMoniker(new EcsResourceService(service));
+
+        if (Objects.equals(currentName.getApp(), moniker.getApp())
+            && Objects.equals(currentName.getDetail(), moniker.getDetail())
+            && Objects.equals(currentName.getStack(), moniker.getStack())) {
+          takenSequences.add(moniker.getSequence());
+        }
+      }
+    }
+
+    // 3. Find the next free sequence number
+    int currentMaxSequence = takenSequences.stream().reduce(Math::max).orElse(0);
+    IntStream after = IntStream.range(currentMaxSequence, SEQUENTIAL_NUMBERING_NAMESPACE_SIZE);
+    IntStream before = IntStream.range(0, currentMaxSequence);
+
+    Moniker.MonikerBuilder nameBuilder =
+        Moniker.builder()
+            .app(currentName.getApp())
+            .detail(currentName.getDetail())
+            .stack(currentName.getStack());
+
+    Moniker newMoniker =
+        IntStream.concat(after, before)
+            .filter(s -> !takenSequences.contains(s))
+            .mapToObj(s -> nameBuilder.sequence(s).build())
+            .limit(MAX_NEXT_SERVER_GROUP_ATTEMPTS)
+            .filter(m -> isNotTaken(m))
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new IllegalArgumentException(
+                        "All server group names for cluster "
+                            + ecsClusterName
+                            + " in "
+                            + region
+                            + " are taken."));
+
+    return new EcsServerGroupName(newMoniker);
+  }
+
+  private boolean isNotTaken(Moniker newMoniker) {
+    String newServiceName = new EcsServerGroupName(newMoniker).getServiceName();
+
+    // An ECS service with this name might exist already in "Draining" state,
+    // so it would not show up in the "taken slots" list.
+    // We need to describe it to determine if it does exist before using the name
+    DescribeServicesRequest request =
+        new DescribeServicesRequest().withCluster(ecsClusterName).withServices(newServiceName);
+    DescribeServicesResult result = ecs.describeServices(request);
+
+    // an active or draining ECS service with this name was not found
+    return result.getServices().isEmpty()
+        || result.getServices().get(0).getStatus().equals("INACTIVE");
+  }
+
+  private List<String> listAllServices(String ecsClusterName) {
+    List<String> allServices = new ArrayList<>();
```

Comment:
nit: can we make this a set instead of a list (to remove duplicates)? `Set<String> allServices = new HashSet<String>();`

Reply:
Maybe, but the calling method relies on being able to chunk the list in to lots of 10 to optimise the calls to `ecs:DescribeServices` efficiently. Changing the return type to a `Set` would make that more complicated. Are you aware of any circumstances where AWS would return duplicate service names from `ecs:ListServices`?

Reply:
It shouldn't but was thinking about it from the perspective that it could possibly happen. That being said, agree with you here, not sure if the effort is necessary.